test/integration/lockfile/test_user_overrides.py

```diff
@@ -360,3 +360,64 @@ def test_lock_update(self, kind, old, new):
         lock = c.load("conan.lock")
         assert old not in lock
         assert new in lock
+
+
+class TestLockUpgrade:
+    @pytest.mark.parametrize("kind, pkg, old, new", [
+        ("requires", "math", "math/1.0", "math/1.1"),
+        # ("build-requires", "cmake", "cmake/1.0", "cmake/1.1"),     # TODO there is not a --build-requires
+        # ("python-requires", "mytool", "mytool/1.0", "mytool/1.1"), # TODO nor a --python-requires
+    ])
+    def test_lock_upgrade(self, kind, pkg, old, new):
+        c = TestClient(light=True)
+        c.save({f"{pkg}/conanfile.py": GenConanfile(pkg)})
+
+        c.run(f"export {pkg} --version=1.0")
+        rev0 = c.exported_recipe_revision()
+        c.run(f"lock create --{kind}={pkg}/[*]")
+        lock = c.load("conan.lock")
+        assert f"{old}#{rev0}" in lock
+
+        c.run(f"export {pkg} --version=1.1")
+        rev1 = c.exported_recipe_revision()
+        c.run(f"lock upgrade --{kind}={pkg}/[*] --update-{kind}={pkg}/[*]")
+        lock = c.load("conan.lock")
+        assert f"{old}#{rev0}" not in lock
+        assert f"{new}#{rev1}" in lock
+
+
+    def test_lock_upgrade_path(self):
+        c = TestClient(light=True)
+        c.save({"liba/conanfile.py": GenConanfile("liba"),
+                "libb/conanfile.py": GenConanfile("libb"),
+                "libc/conanfile.py": GenConanfile("libc")})
+        c.run(f"export liba --version=1.0")
+        c.run(f"export libb --version=1.0")
+        c.run(f"export libc --version=1.0")
+        c.save(
+            {
+                f"conanfile.py": GenConanfile()
+                .with_requires(f"liba/[>=1.0 <2]")
+                .with_requires("libb/[<1.2]")
+                .with_tool_requires("libc/1.0")
+            }
+        )
+
+        c.run("lock create .")
+        lock = c.load("conan.lock")
+        assert "liba/1.0" in lock
+        assert "libb/1.0" in lock
+        assert "libc/1.0" in lock
+
+        c.run(f"export liba --version=1.9")
+        c.run(f"export libb --version=1.1")
+        c.run(f"export libb --version=1.2")
+        c.run(f"export libc --version=1.1")
+        c.run("lock upgrade . --update-requires=liba/* --update-requires=libb/[*] --update-build-requires=libc/[*]")
```

Comment:
It would be great to check that an ``--update-xxx`` doesn't update if the provided version or version range is not in the lockfile, and as such is not removed/unlocked.

Reply:
Yes! This is already being covered in this test case
https://github.com/conan-io/conan/pull/17577/files#diff-1f71b92a9bc5411618c92c12f2976cd7722225a096b9b85a1766d5fe9df533f3R429

Reply:
Not that case. The case when:
- The recipe ``requires()`` contains a version range that would accept the new versions, like ``requires("libb/[*]")``
- The ``conan lock upgrade --update-requires=libb/[>2.0]`` will not upgrade the existing ``libb/1.1``, even if it is in the recipe valid range.